### PR TITLE
[Inference] Dynamic Batching Inference, online and offline

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This PR adds a new feature for colossal inference, dynamic batching, it supports online(with ray serve) and offline (merge inputs into a big batch and returns a generator).

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by ver217** [comment]
Please test these examples: https://github.com/hpcaitech/ColossalAI/tree/main/applications/Colossal-LLaMA-2#examples
And add the screenshots of generation outputs here

File: colossalai/inference/tensor_parallel/modeling/llama.py [hunk 1]
```diff
@@ -363,7 +353,7 @@ def llama_flash_attn_kvcache_forward(
                     infer_state.block_loc,
                     infer_state.start_loc,
                     infer_state.seq_len,
-                    infer_state.cache_manager.past_key_values_length,
+                    infer_state.max_len_in_batch,
                 )
             else:
                 Llama2TokenAttentionForwards.token_attn(
```
Comment (Xu-Kai) [inline colossalai/inference/tensor_parallel/modeling/llama.py:359]: Does this can import from lightllm directly?

**Comment by Xu-Kai** [comment]
The batch State is also used in colossalai/inference/quant/smoothquant/models base_model.py and llama.py. Some variable such as past_key_values_len and cache_manager.past_key_values_len should be updated too.

File: colossalai/inference/tensor_parallel/modeling/llama.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/inference/tensor_parallel/modeling/llama.py:359]: I don't  understand you

**Comment by CjhHa1** [comment]
> Please test these examples: https://github.com/hpcaitech/ColossalAI/tree/main/applications/Colossal-LLaMA-2#examples And add the screenshots of generation outputs here

Sure

File: colossalai/inference/tensor_parallel/modeling/llama.py (hunk 1 above)
Comment (Xu-Kai) [reply colossalai/inference/tensor_parallel/modeling/llama.py:358]: I have checked this function in kernel/triton/token_attention_kernel.py. The function has been already imported from ligthllm directly.

**Comment by CjhHa1** [comment]
image

**Comment by CjhHa1** [comment]
image

**Comment by CjhHa1** [comment]
> Please test these examples: https://github.com/hpcaitech/ColossalAI/tree/main/applications/Colossal-LLaMA-2#examples And add the screenshots of generation outputs here

done

**Review by tiandiao123 — APPROVED** [review]

**Review by ver217 — APPROVED** [review]

**Pull request merged**